rust-src/id/src/ffi.rs

```diff
@@ -292,6 +298,208 @@ pub extern "C" fn ip_info_cdi_verify_key(
+            .expect("Precondition failed: Unable to decode byte array at name_ptr as utf8.")
+            .to_string();
+
+        let url = from_utf8(slice_from_c_bytes!(url_ptr, url_len as usize))
+            .expect("Precondition failed: Unable to decode byte array at url_ptr as utf8.")
+            .to_string();
+
+        let description = from_utf8(slice_from_c_bytes!(desc_ptr, desc_len as usize))
+            .expect("Precondition failed: Unable to decode byte array at desc_ptr as utf8.")
+            .to_string();
+
+        Description {
+            name,
+            url,
+            description,
+        }
+    };
+
+    let ar_info = ArInfo {
+        ar_identity,
+        ar_description,
+        ar_public_key,
+    };
+
+    let bytes = to_bytes(&ar_info);
+    *output_len = to_bytes(&ar_info).len() as size_t;
+    std::mem::forget(bytes);
+    Box::into_raw(Box::new(ar_info))
+}
+
+/// Create a GlobalContext instance from byte array pointers.
+/// Assumes that genesis_string_ptr points to a valid
+/// utf8 string, bulletproof_generators_ptr to a
+/// Generators<G1> instance and on_chain_commitments to a
+/// PedersenKey<G1> instance, and panics otherwise.
+#[no_mangle]
+unsafe extern "C" fn global_context_create(
+    genesis_string_ptr: *const u8,
+    genesis_string_len: size_t,
+    bulletproof_generators_ptr: *const u8,
+    bulletproof_generators_len: size_t,
+    on_chain_commitment_ptr: *const u8,
+    on_chain_commitment_len: size_t,
+    output_len: *mut size_t,
+) -> *mut GlobalContext<G1> {
+    // Genesis string.
+    let genesis_string = from_utf8(slice_from_c_bytes!(
+        genesis_string_ptr,
+        genesis_string_len as usize
+    ))
+    .expect("Precondition failed: Unable to decode byte array at genesis_string_ptr as utf8.")
+    .to_string();
+
+    // Bulletproof generators.
+    let bulletproof_generators = {
+        let bulletproof_generators_buf = &mut slice_from_c_bytes!(
+            bulletproof_generators_ptr,
+            bulletproof_generators_len as usize
+        );
+
+        from_bytes::<Generators<G1>, &[u8]>(bulletproof_generators_buf).expect(
+            "Precondition failed: Unable to create Generators<G1> instance from byte array at \
+             genesis_str_ptr.",
+        )
+    };
+
+    // On-chain commitment key.
+    let on_chain_commitment_key = {
+        let commitment_key_buf =
+            &mut slice_from_c_bytes!(on_chain_commitment_ptr, on_chain_commitment_len as usize);
+
+        from_bytes::<PedersenKey<G1>, &[u8]>(commitment_key_buf).expect(
+            "Precondition failed: Unable to create PedersenKey<G1> instance from byte array at \
+             on_chain_commitment_ptr.",
+        )
+    };
+
+    let global_context = GlobalContext {
+        on_chain_commitment_key,
+        bulletproof_generators,
+        genesis_string,
+    };
+
+    let bytes = to_bytes(&global_context);
+    *output_len = to_bytes(&global_context).len() as size_t;
```

Comment:
```suggestion
    *output_len = bytes.len() as size_t;
```
And please also document the bytes & forgetting here.